spec/datadog/opentelemetry/metrics_spec.rb

```diff
@@ -0,0 +1,373 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+# OpenTelemetry metrics SDK requires Ruby >= 3.1
+if Gem::Version.new(RUBY_VERSION) >= Gem::Version.new('3.1')
+  require 'opentelemetry/sdk'
+  require 'opentelemetry-metrics-sdk'
+  require 'opentelemetry/exporter/otlp_metrics'
+end
+
+require 'datadog/opentelemetry'
+require 'datadog/core/configuration/settings'
+require 'net/http'
+require 'json'
+
+RSpec.describe 'OpenTelemetry Metrics Integration', ruby: '>= 3.1' do
+  let(:default_otlp_http_port) { 4318 }
+
+  before do
+    clear_testagent_metrics
+    Datadog.send(:reset!)
+    provider = ::OpenTelemetry.meter_provider
+    provider.shutdown if provider.is_a?(::OpenTelemetry::SDK::Metrics::MeterProvider)
+    ::OpenTelemetry.meter_provider = ::OpenTelemetry::Internal::ProxyMeterProvider.new
+    allow(Datadog.logger).to receive(:warn)
+    allow(Datadog.logger).to receive(:error)
+    allow(Datadog.logger).to receive(:debug)
+  end
+
+  after do
+    provider = ::OpenTelemetry.meter_provider
+    # Ensures background threads collecting metrics are shutdown.
+    provider.shutdown if provider.is_a?(::OpenTelemetry::SDK::Metrics::MeterProvider)
+  end
+
+  def agent_host
+    Datadog.send(:components).agent_settings.hostname
+  end
+
+  def agent_port
+    Datadog.send(:components).agent_settings.port
+  end
+
+  def clear_testagent_metrics
+    uri = URI("http://#{agent_host}:#{default_otlp_http_port}/test/session/clear")
+    Net::HTTP.post_form(uri, {})
+  rescue => e
+    raise "Error clearing testagent metrics: #{e.class}: #{e}"
+  end
+
+  def get_testagent_metrics
+    uri = URI("http://#{agent_host}:#{default_otlp_http_port}/test/session/metrics")
+
+    try_wait_until(seconds: 2) do
+      response = Net::HTTP.get_response(uri)
+      next unless response.code == '200'
+
+      parsed = JSON.parse(response.body, symbolize_names: false)
+      next parsed if parsed.is_a?(Array) && !parsed.empty?
+    end
+  end
+
+  def find_metric_in_json(payloads, name)
+    payloads.each do |payload|
+      payload['resource_metrics']&.each do |rm|
+        rm['scope_metrics']&.each do |sm|
+          sm['metrics']&.each do |metric|
+            return metric if metric['name'] == name
+          end
+        end
+      end
+    end
+    nil
+  end
+
+  def find_attribute_by_key(attributes, key)
+    attr = attributes&.find { |a| a['key'] == key }
+    attr&.dig('value', 'string_value') || attr&.dig('value', 'int_value') || attr&.dig('value', 'double_value')
+  end
+
+  def setup_metrics(env_overrides = {}, &config_block)
+    ClimateControl.modify({
+      'DD_METRICS_OTEL_ENABLED' => 'true',
+      'DD_AGENT_HOST' => agent_host,
+    }.merge(env_overrides)) do
+      # Reset Datadog to ensure components are reinitialized with the new environment variables
+      Datadog.send(:reset!)
```

Comment:
This `reset!` shouldn't be needed -- we already have one in the before + `Datadog.configure` triggers reconfiguration based on changed env vars already, `reset!` is not needed for that

Reply:
It doesn't return an existing components -- `Datadog.configure { ... }` recreates components, which are then "forced" to pick up the latest configuration (that's exactly why they're recreated).

There's a few corner cases where components get reused (in particular, I believe if you supply your own instances of components then we respect those), but those very rarely come into play (and TBH I'm hoping we can deprecate && remove them at some point as I don't think they're still needed)